### PR TITLE
adds delete note functionality

### LINKED ISSUE #599: Delete ticket note on Frontend
**Purpose**
What will this feature provide?
Ability to delete note in the Ticket modal screen and updates the list in real time.

**Proposed Solution**
If you have an idea on how to implement this feature, please share!

- Add pen/edit button next to the plus sign on ticket
- Clicking the edit button shows a minus/delete icon in the upper righthand corner of existing notes that the user has access to delete.  Currently the assumption will be users can only delete their own notes.
- In the future we might want to give Admins the ability to delete any tickets
- Clicking the minus/delete button causes a "are you sure you want to delete this note?" modal to pop up.  
- If delete is confirmed than call is made to server to delete that note using the id
- When delete is confirmed from server, update local note array to reflect the change
- Clicking the edit button a second time will make the minus/delete buttons disappear
- Closing out the ticket will also make the minus/delete buttons disappear

**Acceptance Criteria**
Please list the requirements the implementation of this feature should meet.
- [] New Edit button displays next to Plus button
- [] Clicking the Edit button will show Minus buttons next to applicable notes
- [] Clicking the Minus button will cause confirmation modal to show
- [] Confirming the confirmation modal will cause the note to be deleted and update the list of notes

**Mockup Screenshot or Page Number**

Paste a screenshot of the feature from the FIGMA mockups or cite the page number.
Screen Shot 2021-04-28 at 7 52 59 PM

**Additional context**
I will create a corresponding backend issue for creating the proper endpoint

### LINKED ISSUE #626: adds edit note functionality to front
**Purpose**

Allows user to edit the text of individual notes on the ticket modal;

Proposed Solution

- Adds "edit pen" to each notes list item on ticket modal.  
- When it is clicked a the text display of the note will turn into a text input field, allowing the user to edit the text of the note. 
-  A small submit and cancel button will also appear in the note list item.  
- When the changes have been made, if the user presses the submit button, a confirmation pop up will appear, confirming the change.  

**Acceptance Criteria**
Please list the requirements the implementation of this feature should meet.
- [ ] Adds edit pen icon next to each note list item
- [ ] When the edit pen is clicked a text input field appears containing current note text
- [ ] When the edit pen is clicked a cancel and submit button appear under text input field 
- [ ] Clicking the submit button triggers confirmation pop up

**Mockup Screenshot or Page Number**
Screen Shot 2021-05-19 at 8 00 50 PM

**Additional context**
I will create a corresponding backend issue for creating the proper endpoint

### PR DESCRIPTION
**Pull request by Protopaco**
### What issue is this solving?
https://github.com/codeforpdx/dwellingly-app/issues/599
Adds "edit pen" to ticket modal, when it is clicked a minus symbol appears next to each note.  If the minus symbol is clicked a confirmation modal appears.  If the delete in confirmed, the note is deleted in the backend, and the local note array is updated.

The "edit pen" will have further use when we add the ability to edit the notes.

### Any helpful knowledge/context for the reviewer?
no

### Please make sure you've attempted to meet the following coding standards
- [x] Code builds successfully
- [x] Code has been tested and does not produce errors
- [x] Code is readable and formatted
- [x] There isn't any unnecessary commented-out code
- [x] There aren't any unnecessary commits or files changed (shouldn't touch the /stashed folder unless the issue requests it)

If you're having trouble meeting this criteria, feel free to reach out on Slack for help!

### CODE REVIEW

**Comment by Protopaco** [comment]
I got a little antsy waiting for the endpoints to be finalized.  These components build on top of each other, so I couldn't build them independently.  I normally like to build things separately and incrementally, but I also want this feature to be done.  So these are still waiting on the backend PRs to be functional.  Thanks!

**Comment by Chris-Boe** [comment]
Looks like the backend was merged, yay!
In the meantime, we did complete some work to switch the Tickets class component to be a functional component, so there are some merge conflicts that might get a little tricky to resolve and then test the changes from this PR again.
@Protopaco let me know if you need any help getting through those, we can tackle it together Wednesday if you like!

**Comment by Protopaco** [comment]
@Chris-Boe I attempted to rebase.  Let me know how I did.  There were a lot of changes, but getting that switched over to a functional component is worth it.  Thanks!

File: src/views/tickets/tickets.jsx [hunk 1]
```diff
@@ -163,6 +165,65 @@ export function Tickets(props) {
...
+
+  const editNote = () => {
+    const ticketID = viewedTicket.id;
+
+    axios.patch(
+      `/api/tickets/${ticketID}/notes/${selectedNote.id}`,
+      { text: selectedNote.text },
+      makeAuthHeaders(userContext))
+      .then(({ data }) => {
+        setViewedTicket({
```
Comment (Chris-Boe) [inline src/views/tickets/tickets.jsx:182]: Let's add a success toast here

File: src/views/tickets/tickets.jsx [hunk 2]
```diff
@@ -163,6 +165,65 @@ export function Tickets(props) {
...
+
+  }
+
+
+  const deleteNote = () => {
+    const { id, ticket_id } = selectedNote;
+
+    axios.delete(`/api/tickets/${ticket_id}/notes/${id}`, makeAuthHeaders(userContext))
+      .then(() => {
+
```
Comment (Chris-Boe) [inline src/views/tickets/tickets.jsx:206]: and another success toast here

File: src/components/NoteListItem/NoteListItem.jsx [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  return (
+    <div className="ticket-card-note-container">
+      <div className="ticket-card-note-header-row">
+        <div className="ticket-card-note-header-row-left">
+          {editNotes ?
+            <div className="ticket-card-note-header-buttons">
+              <button
+                className='mini-rounded'
+                onClick={() => localDeleteNote(note)}>
+                <i className="fas fa-minus icon" />
```
Comment (Chris-Boe) [inline src/components/NoteListItem/NoteListItem.jsx]: Let's switch this out for the "fa-trash" icon

**Review by Chris-Boe — CHANGES_REQUESTED** [review]
Nice job! I tested with the backend 627-edit-note branch and things are looking good. A couple of updates we might want to make while we wait for the Edit notes backend PR to get merged:
* We'll want to add two success Toasts for the user in these workflows, one after successfully deleting a note and one after successfully editing one.
* We should update the delete icon from a minus sign to a trash icon to be consistent with our other delete workflows
* The text seems to be left aligned when it takes up only one line, but as soon as you add more lines of text it switches to right-aligned. Do we want all text to be left aligned or right aligned? I've also done an implementation where the viewing user's notes were right aligned while other notes (from other users) were left aligned, but we probably don't want to make it too much like a chat system. We should be consistent though so maybe switch everything to left aligned?
image

File: src/components/NoteListItem/NoteListItem.jsx (hunk 3 above)
Comment (Protopaco) [reply src/components/NoteListItem/NoteListItem.jsx]: Makes perfect sense.

**Comment by Protopaco** [comment]
Thanks for catching these changes.  I hadn't even noticed that text align bug.  I believe they are all addressed in that last commit.  

One other thing that just popped into my mind, do we want to put a character limit on the note creation?  The design is responsive, so that shouldn't be an issue, but it would be an easy limit to add.

Thanks!

**Comment by Chris-Boe** [comment]
> Thanks for catching these changes. I hadn't even noticed that text align bug. I believe they are all addressed in that last commit.
> 
> One other thing that just popped into my mind, do we want to put a character limit on the note creation? The design is responsive, so that shouldn't be an issue, but it would be an easy limit to add.
> 
> Thanks!

We could put a higher limit on it if we need to, like 400 characters, but we might not need it

**Comment by Protopaco** [comment]
> We could put a higher limit on it if we need to, like 400 characters, but we might not need it

Sounds good, no need to add restrictions if they aren't needed.  It will be an easy fix later if we need to.

**Review by Chris-Boe — APPROVED** [review]
@Protopaco Sorry I hadn't looked at this for a while! I went ahead and fixed the merge conflicts from the other PR's we got in. Everything looks great!

**Pull request merged**